src/loop_logic.rs

```diff
@@ -315,8 +377,29 @@ impl<'l, Data> EventLoop<'l, Data> {
                 };
             }
         };
+        {
+            let mut extra_lifecycle_sources = self
+                .handle
+                .inner
+                .sources_with_additional_lifecycle_events
+                .borrow_mut();
+            if !extra_lifecycle_sources.values.is_empty() {
+                for (source, has_event) in &mut *extra_lifecycle_sources.values {
+                    for event in &events {
+                        *has_event |= (event.token.key & MAX_SOURCES_MASK) == source.key;
+                    }
+                }
+            }
+            for (source, has_event) in &*extra_lifecycle_sources.values {
+                if let Some(disp) = self.handle.inner.sources.borrow().get(source.key) {
+                    disp.before_handle_events(*has_event);
+                } else {
+                    unreachable!()
+                }
+            }
```

Comment:
Couldn't these two loops be fused into a single one?

And then, it looks like `has_event` is only used by those two loops, so it can probably be removed altogether from `extra_lifecycle_sources`, no?

Reply:
I don't believe so - we need to run all before_handle_events handlers, even those which didn't receive events

I suppose we could have two different handler methods for the different parts, so that we don't need to track was_awoken. That means that if a source needs was_awoken, it would have to track it itself. But that's fine

I'm fairly sure we do need two loops though

Reply:
But what about something like this?

```
for each extra_lifecycle_source {
    let has_event = /* loop over events to check if that source was woken up */;
    disp.before_handle_events(has_event);
}